### ISSUE TITLE
Cryptomator storage not possible

### ISSUE THREAD

**Issue by git4eb7** [keepassxreboot/keepassxc#790]
Cryptomator can be used to create a secure enclave inside e.g. a cloud storage and is a free competitor against other cloud storgae encryption products. It is available on all major platforms.

## Expected Behavior
It should be possible to use Cryptomator instead of other cloud storage enryption products like boxcryptor (which works fine as it uses a diffent way of integration).

## Current Behavior
While importing a kdb1 file it is not possible to store the new database into the WebDAV folder of Cryptomator. 

## Possible Solution
Support Cryptomator directly using the WebDAV interface or just allow to write onto such a volume.

## Steps to Reproduce (for bugs)
1. import a kdbv1 file and next try to save it to a cryptomator volume.It does not matter whether the storage is local or on cloud-storage.

## Context
I use Cryptomator as a platform independent cloud storage encryption provider between Linux, MacOS Adroid and iOS. The goal is to have a security enclave inside of the cloud storage for files which need additional security like my kdb file(s).

## Debug Info
KeePassXC - 2.2 Windows 32bit

Libraries:
- default

Operating system: Windows
CPU architecture: x86 
Kernel: Win7 Pro

Enabled extensions:
- default

**Comment by phoerious** [comment]
I don't understand what you are trying to do. The KDBX file is already encrypted.

**Comment by git4eb7** [comment]
Hi,

yes of course is the kdb file already encrypted. I just want to add another
layer of security which is very easy to do with cloud storage encryption
providers and makes the life of hackers crackers and three letter agencies
really hard.

Best,
Robert

On Tue, 18 Jul 2017 13:24:33 +0000 (UTC), Janek Bevendorff wrote

**Comment by phoerious** [comment]
Sorry, but I don't really see the need for this. You can already increase the security of your database by using a strong master key, adjusting the number of transformation rounds, using a key file (which is not synced with the DB) or a YubiKey.

Adding support for various different boxcryptor formats would add a large amount of complexity to KeePassXC without much benefit.

**Comment by git4eb7** [comment]
Hi,

I'm working as a principal security specialist with tons of security
certifications including CISSP and others and you can believe me it adds a
massive value to keepass if you support any storage which is transparent.

The complexity for keepassXC will remain the same as it is today. Cloud
Storage encryption is fully transparent, its just a local filesystem.

It would have been enough to diable the error message while trying to write to
a webdav filespace. Thats all. So this would reduce complexity instead of
adding to KeepassXC. 

Obviously you are not familar with security models and I suggest to you
personally to play with security models. This will help a lot as it changes
your point of view dramatically and adds a big value as the world is currenty
changing rapidly.

Please have a look at the model
https://en.wikipedia.org/wiki/Defense_in_depth_(computing)

It was just a comment and it blocks at least me of making use of your port.

Best,
Robert
  
On Tue, 18 Jul 2017 06:36:03 -0700, Janek Bevendorff wrote
> Sorry, but I don't really see the need for this. You can already 
> increase the security of your database by using a strong master key, 
> adjusting the number of transformation rounds, using a key file 
> (which is not synced with the DB) or a YubiKey.
> 
> Adding support for various different boxcryptor formats would add a 
> large amount of complexity to KeePassXC without much benefit.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/keepassxreboot/keepassxc/issues/790#issuecomment-316065576


Best,
Robert
--  
ThreemaID: 2C7Z4J6V

**Comment by phoerious** [comment]
I know what you are trying to do, but from your description it sounds like you want us to support specific cloud providers (or cloud encryption layers). We have had this discussion quite a few times and are still undecided if we want to natively support any network / cloud synchronization at all for the reasons above.

Now your last comment draws a different picture. You already have your WebDAV drive mounted locally and want KeePassXC to be able to write to it. That should already be possible. If not, it's a bug and you should maybe rephrase your issue report and give us more details.

**Comment by git4eb7** [comment]
Hi,

thats true and I fully agree not to implement any things from outside which
could change quickly and therefor will break keepassxc. 
This approach is not only valid but will grant roubustness as well. 
Please do NOT integrate any cloud-providers. This wont work on all platforms
and adds a lot of compexity. 
Dropbox and Co can do this automagically and I use this now for years with
little inconveniances and a learning curve what to do and what not. 

So I'm happy to get this sorted. Could you please go back to the initial
state, where all the details I have are already stated.

Its really just the errormessage that pops up when I try to save the new file
to the cryptomator space. 
The messagebox screenshot is attached as jpg.

Best,
Robert

On Tue, 18 Jul 2017 14:22:28 +0000 (UTC), Janek Bevendorff wrote
> I know what you are trying to do, but from your description it 
> sounds like you want us to support specific cloud providers (or 
> cloud encryption layers). We have had this discussion quite a few 
> times and are still undecided if we want to natively support any 
> network / cloud synchronization at all for the reasons above.
> 
> Now your last comment draws a different picture. You already have 
> your WebDAV drive mounted locally and want KeePassXC to be able to 
> write to it. That should already be possible. If not, it's a bug and 
> you should maybe rephrase your issue report and give us more details.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/keepassxreboot/keepassxc/issues/790#issuecomment-316078462


Best,
Robert
--  
ThreemaID: 2C7Z4J6V

**Comment by phoerious** [comment]
We actually encourage people to use transparent cloud storage, but we have also heard a lot of concerns about update safety when multiple people use the same database and Dropbox on windows seems to lock the file resulting in write errors (maybe we have the same issue here?). Such problems are hard to fix without implementing cloud service support directly inside KeePassXC (which we are still reluctant to do).

I don't see any screenshot. I suppose GitHub doesn't support mail attachments. Could you please go ahead and edit your original post to include it there? Thanks!

**Comment by git4eb7** [comment]
Here the screenshot of the popup when trying to save the new database to a cryptomator volume on Windows:



![errmsgwriteoncryptom](https://user-images.githubusercontent.com/16333225/28325020-1e8c20ee-6bdd-11e7-99af-d250fc858de9.jpg)

**Comment by git4eb7** [comment]
Thanks fpr reopening - was just by accident

**Comment by phoerious** [comment]
Is Y:\ a network drive or is it a folder under Y:\?

**Comment by git4eb7** [comment]
Its a network drive as Cryptomator uses WebDAV and assigned it to drive letter Y: It is linked to \\localhost@42427\3kfds2piq0 what looks a bit cryptic to me.

**Comment by phoerious** [comment]
I suppose Windows network drives are not as transparent as they should be.

**Comment by git4eb7** [comment]
yes - it works well with Boxcryptor which assigns a drive letter as well.

**Comment by git4eb7** [comment]
I've tried the same procedure with Mac OS/X 10.11 and Cryptomator and here it works well. No problem at all. So its definitely a Windows Client problem.

**Comment by git4eb7** [comment]
With my KALI2 (current patches all applied) its not working either. In Linux Cryptomator presents a folder named: dav://localhost:654345/Goohjkhkh/myCryptomatorTest. I can see the folder as network drive in the filebrowser but its not offered in the save-as dialog with KeepassXC.

**Comment by git4eb7** [comment]
With a little trick using davfs2 and mounting the url provided by cryptomator it works well and the storage can be accessed from the save-as dialog from within KeepassXC. So the "Linux" part is solved as well.

**Comment by phoerious** [comment]
On Kali, are you using KDE or Gnome? If so, they use their own file system abstraction layer which only works with applications that directly support it. You need to mount the folder on file system level (as you've figured out).

**Comment by droidmonkey** [comment]
For the record, encrypting already encrypted data does not necessarily make it more secure than just increasing the length of your key.

Defense in depth refers to having multiple layers of protection in a network, not encryption. It involves using firewalls, virtual lans, hbss, and other technologies to make it much harder to get to the juicy center.

Just wanted to clear that all up from the CISSP ramblings above.

**Comment by git4eb7** [comment]
Yes the cryptomator drive y: works with the office suite as well as with the native Windows programms like wordpad.

**Comment by phoerious** [comment]
I've had problems with network drive with various applications in the past. They are not as transparent as, say, an NFS drive mounted to a *nix machine.

**Comment by OmanF** [comment]
@git4eb7 could you explain/link/(any other resource) how you used the davfs2 trick.
I'm on Antergos and am having some issues mounting the sucker in a read-**write** fashion.

Thanks.
